### DISCUSSION TITLE
How to deal with client/subcription inactivity?

### DISCUSSION THREAD

**Discussion by ErikTempelaarVO** [open62541pp/open62541pp#284] [Q&A] [answered]
I am testing against a misbehaving OPC UA server that becomes unresponsive.
(Running in a docker container that I simply 'pause')

For notification I have the following callbacks setup, as below:
- inactivityCallback
- subInactivityCallback

```   // Verbose logging callback for subscription inactivity
  auto subInactivityCallback = [](UA_Client* client, UA_UInt32 subId, void* subscriptionContext) {
    (void)subscriptionContext;
    const UA_ClientConfig* cc = UA_Client_getConfig(client);
    UA_LOG_INFO(&cc->logger, UA_LOGCATEGORY_USERLAND, "Inactivity for subscription %u", subId);
  };

  // Verbose logging callback for client inactivity
  auto inactivityCallback = [](UA_Client* client) {
    const UA_ClientConfig* cc = UA_Client_getConfig(client);
    UA_LOG_INFO(&cc->logger, UA_LOGCATEGORY_USERLAND, "Inactivity for client");
  };
  auto cc = UA_Client_getConfig(m_client.handle());
  cc->connectivityCheckInterval = 10000;
  cc->subscriptionInactivityCallback = subInactivityCallback;
  cc->inactivityCallback = inactivityCallback;
``` 
The open62541 [documentation ](https://www.open62541.org/doc/1.3/client.html#advanced-client-configuration)hints to restarting the connecting as an atttempt to fix the broken connection, how to do that in open62541pp?
I don't think the passed contexts (subcriptioncontext/clientconfig context) contain anything I should be using directly?
I'd like to do some bookkeeping that the connection is failing.

**Comment by ErikTempelaarVO** [comment]
Any thoughts?

**Reply by lukasberbuer** [reply]
Hi @ErikTempelaarVO, sorry for the late reply.
If I understand you correctly, you would like to gain access to the `Client` instance through open62541's inactivity callback, where just the `UA_Client` instance and the context pointers are passed?

**Reply by ErikTempelaarVO** [reply]
Thanks for following up on this question.

I'd like way to handle these events in the class that owns the opcua::Client, aside from logging I'd like to mark the connection as invalid and schedule a reconnection attempt / restart the subscription?

Maybe similar to the other callbacks onSessionActivated / onSessionClosed / onDisconnected etc.? Is that feasible?
```
  m_client.onDisconnected([this, title]() {
    spdlog::info("{} has been disconnected", title);
    handleClientDisconnected(); // calls member of this to do some bookkeeping
  });
```

Hope that helps to understand what I'm after.

**Comment by lukasberbuer** [comment, marked as answer]
Ok, this is was pretty easy to implement. Can you review and try solution in PR #304 please?

**Reply by ErikTempelaarVO** [reply]
Solution works like a charm. Thanks.
